src/lesson_2/Vault.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.12;
+
+import "yield-utils-v2/contracts/token/IERC20.sol";
+
+/**
+@title A basic vault
+@author Sabnock01
+@notice This contract allows for deposits and withdrawals 
+of an ERC20 token to a vault
+*/
+contract Vault {
+    /// @notice the ERC20 token the vault holds
+    IERC20 public token;
+
+    /// @notice mapping from user addresses to their respective vault balance
+    mapping(address => uint256) public balances;
+
+    /// @notice  emitted when the user deposits to the vault
+    event Deposited(address indexed from, uint amount);
+    /// @notice emitted when the user withdraws from the vault
+    event Withdrawn(address indexed to, uint amount);
+
+    /// @param token_ the address of the ERC20 token to instantiate
+    constructor(IERC20 token_) {
+        token = token_;
+    }
+
+    /**
+    @notice Deposits to the vault
+    @param amount the amount to deposit
+     */
+    function deposit(uint256 amount) public {
+        balances[msg.sender] += amount;
```

Comment:
Could you use `unchecked` here?

Reply:
Like this? `unchecked { balances[msg.sender] += amount; }`

Reply:
Yes, but the question is, would it be safe?

Reply:
Not unless it's doing bounds checking some other place which I don't *think* it is.

Reply:
It would actually be safe, as long as the deposited token is known to be not malicious.

The deposited balance of any account can't ever be more than the total supply of the deposited token. No one can own more than `type(uint256).max` of the underlying token.

I'm being a bit pedantic in here, but I wanted to point out that sometimes conditions are enforced by construction, instead of being enforced by `require` statements.

Reply:
I understand.